connector/google/google.go

```diff
@@ -0,0 +1,192 @@
+	verifier      *oidc.IDTokenVerifier
+	ctx           context.Context
+	cancel        context.CancelFunc
+	logger        logrus.FieldLogger
+	hostedDomains []string
+	serviceAccountCred  string
+	subject         string
+	groupPrefix	string
+}
+func (c *googleConnector) Close() error {
+	c.cancel()
+	return nil
+}
+func (c *googleConnector) LoginURL(s connector.Scopes, callbackURL, state string) (string, error) {
+	if c.redirectURI != callbackURL {
+		return "", fmt.Errorf("expected callback URL %q did not match the URL in the config %q", callbackURL, c.redirectURI)
+	}
+	if len(c.hostedDomains) > 0 {
+		preferredDomain := c.hostedDomains[0]
+		if len(c.hostedDomains) > 1 {
+			preferredDomain = "*"
+		}
+		return c.oauth2Config.AuthCodeURL(state, oauth2.SetAuthURLParam("hd", preferredDomain)), nil
+	}
+	return c.oauth2Config.AuthCodeURL(state), nil
+}
+type oauth2Error struct {
+	error            string
+	errorDescription string
+}
+func (e *oauth2Error) Error() string {
+	if e.errorDescription == "" {
+		return e.error
+	}
+	return e.error + ": " + e.errorDescription
+}
+func (c *googleConnector) HandleCallback(s connector.Scopes, r *http.Request) (identity connector.Identity, err error) {
+	q := r.URL.Query()
+	if errType := q.Get("error"); errType != "" {
+		return identity, &oauth2Error{errType, q.Get("error_description")}
+	}
+	token, err := c.oauth2Config.Exchange(r.Context(), q.Get("code"))
+	if err != nil {
+		return identity, fmt.Errorf("oidc: failed to get token: %v", err)
+	}
+	rawIDToken, ok := token.Extra("id_token").(string)
+	if !ok {
+		return identity, errors.New("oidc: no id_token in token response")
+	}
+	idToken, err := c.verifier.Verify(r.Context(), rawIDToken)
+	if err != nil {
+		return identity, fmt.Errorf("oidc: failed to verify ID Token: %v", err)
+	}
+	var claims struct {
+		Username      string `json:"name"`
+		Email         string `json:"email"`
+		EmailVerified bool   `json:"email_verified"`
+		HostedDomain  string `json:"hd"`
+	}
+	if err := idToken.Claims(&claims); err != nil {
+		return identity, fmt.Errorf("oidc: failed to decode claims: %v", err)
+	}
+	if len(c.hostedDomains) > 0 {
+		found := false
+		for _, domain := range c.hostedDomains {
+			if claims.HostedDomain == domain {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return identity, fmt.Errorf("oidc: unexpected hd claim %v", claims.HostedDomain)
+		}
+	}
+	identity = connector.Identity{
+		UserID:        idToken.Subject,
+		Username:      claims.Username,
+		Email:         claims.Email,
+		EmailVerified: claims.EmailVerified,
+	}
+	if s.Groups {
+		groups, _ := c.getGroups(identity.Email)
+		if len(groups) == 0{
+			return identity, fmt.Errorf("oidc: no groups for user %v", identity.Email)
+		}
+		identity.Groups = groups
+	}
+	return identity, nil
+}
+// getGroups retrieves groups starts with the prefix groupPrefix
+func (c *googleConnector) getGroups(userLogin string) (groups []string, err error) {
+	b := []byte(c.serviceAccountCred)
```

Comment:
this logic for setting up the admin directory api client should not happen with every request. we should add a field to the googleConnector struct and initialize it during the call to the open function above. This logic should also get extracted into a method on the config object above.